### PR TITLE
MONGOID-5161 - [Documentation Only] Clarify the policy for working with . and $ named fields which were introduced in MongoDB 5.0

### PR DESCRIPTION
**Pull request by johnnyshields**
**HELP: I cannot create tickets in MongoDB JIRA. When I open "Create Ticket" then press submit, the JIRA server returns 200 but no ticket is made. (I've tried in multiple browsers and with multiple JIRA accounts.) In order to follow ticket process as best as possible, until the issue is resolved I will:**
* **mark Github PRs with `MONGOID-????`**
* **include this message in each affected PR**

---

Documentation only.

In summary:
* `$` and `.` fields (introduced in MongoDB 5.0) are special "hidden fields" that can only be accessed with special "shadow" operators.
* Mongoid query operators / field names will behave as normal post-MongoDB 5.0
* The `field` function will not allow `.` and `$` in field names; Ruby doesn't support such method names anyway.

### CODE REVIEW

**Comment by p-mongo** [comment]
@johnnyshields can you try creating an issue from the form at https://jira.mongodb.org/secure/CreateIssue!default.jspa ? I escalated this internally as well.

**Comment by johnnyshields** [comment]
@p-mongo I really appreciate the help. Here's what I see when I try to create an issue by the link you posted:

image

Can you see https://jira.mongodb.org/browse/MONGOID-5161 ? Here's what I see:

image

File: docs/reference/fields.txt [hunk 1]
```diff
@@ -1079,3 +1081,70 @@ you can include the short versions of the modules.
...
+Support in Mongoid
+------------------
+
+This section outlines the behavior of Mongoid and the Mongo Ruby Driver
+when working with fields whose names are prefixed with ``$`` or contain ``.``
+("specially-named fields"):
+
+* Mongoid's :ref:`field <field-types>` macro does not support defining specially-named fields.
+* Mongoid does not support getting or setting the values of specially-named fields.
+* Mongoid does not support queries or operations which reference specially-named fields.
```
Comment (p-mongo) [inline docs/reference/fields.txt:1112]: I don't see how this is true. If you are using 5.0 server you should be able to query these fields using the respective server operators.

File: docs/reference/fields.txt [hunk 2]
```diff
@@ -1079,3 +1081,70 @@ you can include the short versions of the modules.
...
+
+.. code-block:: ruby
+
+  Person.where('www.mongodb.com' => 'favorite_website').to_a
+  #=> []
+  # queries for `www: { mongodb: { com: 'favorite_website' } }`
+
+Lastly, please note that method names containing ``$`` or ``.``
+are `not permitted in Ruby language
+<https://docs.ruby-lang.org/en/3.0.0/doc/syntax/methods_rdoc.html>`_.
```
Comment (p-mongo) [inline docs/reference/fields.txt:1150]: This is also not true:

```
define_method('foo.bar') { 42 }
send('foo.bar')
# => 42
```

We actually had a ticket filed by someone a while ago who generated method names with spaces in them based on MongoDB field names.

**Review by p-mongo — COMMENTED** [review]
I responded to the core idea of this proposal in https://jira.mongodb.org/browse/MONGOID-5161. The inline comments are reactions to specific changes that are proposed here.

File: docs/reference/fields.txt [hunk 3]
```diff
@@ -1079,3 +1081,70 @@ you can include the short versions of the modules.
...
+  #=> raises Mongo::Error::OperationFailure (unknown top level operator: $my_money)
+
+In both MongoDB and Mongoid, ``.`` is reserved for embedded field references, and will return
+an empty result:
+
+.. code-block:: ruby
+
+  Person.where('www.mongodb.com' => 'favorite_website').to_a
+  #=> []
+  # queries for `www: { mongodb: { com: 'favorite_website' } }`
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply docs/reference/fields.txt:1146]: Have clarified my meaning:
```
queries for a nested document such as `www: { mongodb: { com: 'favorite_website' } }`
```

File: docs/reference/fields.txt (hunk 2 above)
Comment (johnnyshields) [reply docs/reference/fields.txt:1150]: That is hideous. And no, that doesn't count as a standard method because I cannot define it with `def` and call it as `object.foo.bar`.

I pray to the good lord that you are are not considering supporting `.` / `$` in this manner.

File: docs/reference/fields.txt [hunk 4]
```diff
@@ -1079,3 +1081,70 @@ you can include the short versions of the modules.
...
+to be indexed.
+
+Support in Mongoid
+------------------
+
+This section outlines the behavior of Mongoid and the Mongo Ruby Driver
+when working with fields whose names are prefixed with ``$`` or contain ``.``
+("specially-named fields"):
+
+* Mongoid's :ref:`field <field-types>` macro does not support defining specially-named fields.
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply docs/reference/fields.txt:1110]: This is horrendous.

File: docs/reference/fields.txt [hunk 5]
```diff
@@ -1079,3 +1081,70 @@ you can include the short versions of the modules.
...
+* Mongoid does not support getting or setting the values of specially-named fields.
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply docs/reference/fields.txt:1111]: It should **not** be possible with `$set`, because MongoDB's `$set` does not allow it. You must use the `$setField` operator instead.

Please, please do not hack `$set` in Mongoid so that it magically calls `$setField` under the hood. Instead, if we need to, make `$setField` it's own special atomic operator.

File: docs/reference/fields.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnnyshields) [reply docs/reference/fields.txt:1110]: Yes, but it doesn't support saving to the database today, does it?

File: docs/reference/fields.txt (hunk 1 above)
Comment (johnnyshields) [reply docs/reference/fields.txt:1112]: Does Mongoid support this today?

File: docs/reference/fields.txt [hunk 6]
```diff
@@ -1079,3 +1081,70 @@ you can include the short versions of the modules.
...
+* Mongoid does not support queries or operations which reference specially-named fields.
+* If a document existing in the database contains specially-named fields,
+  Mongoid will load the document normally **except** for its specially-named fields,
+  which will not be retrieved.
```
[the remark this answers is not in the document]
Comment (johnnyshields) [reply docs/reference/fields.txt:1115]: Did you test this on MongoDB 5.0? I do not believe the DB will return them on standard queries.

**Comment by p-mongo** [comment]
The purpose of my examples was to demonstrate the various ways that someone can *already* use to work with data which happens to include dots and dollars. It is a known fact that dots and dollars are treated specially by various MongoDB-related software and therefore they require care, and aren't easy to use in general in field names. However I don't see how stating that X is "not supported" or "not permitted" when there are ways of achieving it in some capacity is helpful.

Especially given that the direction of MongoDB is to make dots and dollars in field names more usable, adding documentation stating this sort of thing is not supported/not permitted isn't aligned with the direction in which MongoDB is heading.

**Comment by johnnyshields** [comment]
> the direction of MongoDB is to make dots and dollars in field names more usable

A better way to say this is "MongoDB is making dots and dollars *possible* to use, with **extreme caveats, drawbacks, and limitations.**" Such fields will not be treated as "first-class citizens"; you cannot index them, you use special operators to get/set them, etc. This is clear in the [MongoDB documentation](https://docs.mongodb.com/manual/core/dot-dollar-considerations/#std-label-crud-concepts-dot-dollar-considerations). They are only being supported for edge-cases of working with legacy data-sources which for whatever reason can't be mapped.

Mongoid/Ruby should take a similar approach: In the future, it should be *possible* to access such fields, but it should not should not be treated as "standard", an no path traversal/operator functionality should be sacrificed in order to support them.

File: docs/reference/fields.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (p-mongo) [reply docs/reference/fields.txt:1110]: This is a driver limitation. Once it is lifted in the driver, I expect all published Mongoid versions will be able to persist fields like this.

File: docs/reference/fields.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (p-mongo) [reply docs/reference/fields.txt:1115]: I believe I tested on 5.0 but if you wish I can test again.

File: docs/reference/fields.txt (hunk 1 above)
Comment (p-mongo) [reply docs/reference/fields.txt:1112]: What I meant was that you should be able to use the 5.0 server operators via the MQL syntax. Mongoid supports this (I would say) because it passes through MQL syntax to the driver and ultimately the server.

Mongoid doesn't presently provide *helpers* for the new 5.0 query operators, but a user utilizing Mongoid should be able to perform these queries hence I'd rather not state that something is not possible when there is a way to do it.

File: docs/reference/fields.txt (hunk 2 above)
Comment (p-mongo) [reply docs/reference/fields.txt:1150]: At this point I am not aware of any plans for dot or dollar changes in Mongoid. In any event I would prefer to not break compatibility, hence I think we are largely on the same page with respect to Mongoid's direction but again I am unaware of any specific changes that are under consideration for Mongoid right now.

**Comment by p-mongo** [comment]
I intend to propose some language here and I'll put this PR into my todo list to remind me to do so.

@johnnyshields  would you prefer me to add commits to this PR with changes, make my own PR or something else?

File: docs/reference/fields.txt (hunk 4 above)
[the remark this answers is not in the document]
Comment (johnnyshields) [reply docs/reference/fields.txt:1110]: @p-mongo from your comments on this issue it appears you are fundamentally misunderstanding how these fields work in MongoDB, and furthermore you have a mistaken conception that they can be treated like regular fields in Mongoid to provide a "seamless" experience to the end-user.

Although such an approach may be **possible** to some extent, it will require so many hacks and caveats that will not be **practical**. The fields should be handled specially and only accessed via specialized `get`/set` functions.

**Comment by johnnyshields** [comment]
@p-mongo it's fine to commit the changes to this branch.

File: docs/reference/fields.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (johnnyshields) [reply docs/reference/fields.txt:1115]: Yes, please provide evidence of this. The MongoDB documentation seems to indicate such fields won't be returned.

File: docs/reference/fields.txt (hunk 1 above)
Comment (johnnyshields) [reply docs/reference/fields.txt]: I've clarified the language "Mongoid does not provide helper functions or idiomatic support..."

**Comment by johnnyshields** [comment]
@p-mongo any update on this?

**Review by p-mongo — APPROVED** [review]

**Review by comandeo — APPROVED** [review]
